util/certs/certs.go

```diff
@@ -113,3 +118,73 @@ func DecodePrivateKeyPEM(encoded []byte) (crypto.Signer, error) {
 
 	return nil, kerrors.NewAggregate(errs)
 }
+
+// NewSigner creates a private key based on the provided encryption key algorithm.
+func NewSigner(keyEncryptionAlgorithm bootstrapv1.EncryptionAlgorithmType) (crypto.Signer, error) {
+	switch keyEncryptionAlgorithm {
+	case bootstrapv1.EncryptionAlgorithmECDSAP256:
+		return ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	case bootstrapv1.EncryptionAlgorithmECDSAP384:
+		return ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
+	}
+	rsaKeySize := rsaKeySizeFromAlgorithmType(keyEncryptionAlgorithm)
+	if rsaKeySize == 0 {
+		return nil, errors.Errorf("cannot obtain key size from unknown RSA algorithm: %q", keyEncryptionAlgorithm)
+	}
+	pk, err := rsa.GenerateKey(rand.Reader, rsaKeySize)
+	return pk, errors.WithStack(err)
+}
+
+// EncodePrivateKeyPEMFromSigner converts a known private key type of RSA or ECDSA to
+// a PEM encoded block or returns an error.
+func EncodePrivateKeyPEMFromSigner(key crypto.PrivateKey) ([]byte, error) {
+	switch t := key.(type) {
+	case *ecdsa.PrivateKey:
+		derBytes, err := x509.MarshalECPrivateKey(t)
+		if err != nil {
+			return nil, err
+		}
+		block := &pem.Block{
+			Type:  "EC PRIVATE KEY",
+			Bytes: derBytes,
+		}
+		return pem.EncodeToMemory(block), nil
+	case *rsa.PrivateKey:
+		block := &pem.Block{
+			Type:  "RSA PRIVATE KEY",
+			Bytes: x509.MarshalPKCS1PrivateKey(t),
+		}
+		return pem.EncodeToMemory(block), nil
+	default:
+		return nil, fmt.Errorf("private key is not a recognized type: %T", key)
+	}
+}
+
+// EncodePublicKeyPEMFromSigner returns PEM-encoded public key data.
+func EncodePublicKeyPEMFromSigner(key crypto.PublicKey) ([]byte, error) {
+	der, err := x509.MarshalPKIXPublicKey(key)
+	if err != nil {
+		return []byte{}, errors.WithStack(err)
+	}
+	block := pem.Block{
+		Type:  "PUBLIC KEY",
+		Bytes: der,
+	}
+	return pem.EncodeToMemory(&block), nil
+}
+
+// rsaKeySizeFromAlgorithmType takes a known RSA algorithm defined in the kubeadm API
+// and returns its key size. For unknown types it returns 0. For an empty type it returns
+// the default size of 2048.
+func rsaKeySizeFromAlgorithmType(keyEncryptionAlgorithm bootstrapv1.EncryptionAlgorithmType) int {
+	switch keyEncryptionAlgorithm {
```

Comment:
Let's add a godoc here that "" is the default on the API field and we default that here to RSA 2048